packages/compiler/core/program.ts

```diff
@@ -204,6 +204,7 @@ export async function createProgram(
   const requireImports = new Map<string, string>();
   const libraryLoaded = new Set<string>();
   let error = false;
+  let parseAndBindCache: WeakMap<SourceFile, CadlScriptNode> | undefined;
```

Comment:
And... This doesn't actually work since there's a new cache per program. 🤦‍♂️ It would need to be global. I think I'd be more comfortable if it were associated with the host object instead. Thinking about it. Open to suggestions.

Reply:
Moved the cache to the host.

Reply:
wondering if we'd ever want a more generic cache in the future. Cross program caching. The way typescript does it is also to take `oldProgram` as one of the program argument.

Reply:
Interesting. Yes, I think we will eventually want more. It will be a bit messy to keep track of oldProgram in the server but I'll take a look.

Reply:
I did this and it feels cleaner and there is now opportunity to do smarter things with oldProgram. The first such smart thing I did was to stop using parse() in the LS and use compile() instead, and have the compiler return oldProgram unchanged if the complete set of files are the same after loading all imports. This way, things like folding/semantic highlighting, etc. can reuse the same compilation as checking the document for errors, rename, and other things that need a full compilation. :)

The one theoretical downside is that we will not share parse treee between different entry points because they won't have the same oldProgram. I could fix this with a hybrid of both approaches, but I thought it wasn't worth the complexity.

Reply:
Updated description. Still needs testing and perf measurement, but I think it's promising. I traced when we were reusing things and it's A LOT. :)

Reply:
Some more thoughts/notes to self:

* I probably do want to also reuse parsing across entry points because workspace symbols when added would parse every document and it would be great if these don't get re-parsed for other features. This could wait though.

* Probably need to check more than source file set when skipping checking and returning oldProgram. What if options are different?

Reply:
* Trade-off that parse-only features repeat search for main file. Measure impact.

Reply:
Fixed all of my notes to self here. Turns out the I/O is still significant poking around for the main file and resolving imports, etc. So it is worthwile to skip this in features that only need to parse. So now I have both the parseCache and the oldProgram.

I think we can do better on avoiding this I/O and I will file an issue.

Reply:
Decription updated. I'm still thinking about the best way to test this.